File: analytics/src/test/java/io/appmetrica/analytics/impl/referrer/service/ReferrerValidityCheckerDoesInstallerMatchReferrerTest.java
package io.appmetrica.analytics.impl.referrer.service;

import android.content.Context;
import androidx.annotation.NonNull;
import io.appmetrica.analytics.coreutils.internal.services.SafePackageManager;
import io.appmetrica.analytics.impl.IReporterExtended;
import io.appmetrica.analytics.impl.referrer.common.ReferrerInfo;
import io.appmetrica.analytics.testutils.CommonTest;
import io.appmetrica.analytics.testutils.TestUtils;
import java.util.Arrays;
import java.util.Collection;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.ParameterizedRobolectricTestRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

@RunWith(ParameterizedRobolectricTestRunner.class)
public class ReferrerValidityCheckerDoesInstallerMatchReferrerTest extends CommonTest {

    @ParameterizedRobolectricTestRunner.Parameters(name = "{0}")
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][]{
            {ReferrerInfo.Source.UNKNOWN, false, false},
            {ReferrerInfo.Source.HMS, false, true},
            {ReferrerInfo.Source.GP, true, false},
        });
    }

    private static final String GOOGLE_INSTALLER = "com.android.vending";
    private static final String HUAWEI_INSTALLER = "com.huawei.appmarket";

    private Context context;
    @Mock
    private SafePackageManager packageManager;
    @Mock
    private IReporterExtended selfReporter;
    private final String packageName = "ru.yandex.test";
    @NonNull
    private final ReferrerInfo referrerInfo;
    private final boolean matchesGoogle;
    private final boolean machesHuawei;
    private ReferrerValidityChecker referrerValidityChecker;

    public ReferrerValidityCheckerDoesInstallerMatchReferrerTest(@NonNull ReferrerInfo.Source source,
                                                                 boolean matchesGoogle,
                                                                 boolean matchesHuawei) {
        this.referrerInfo = new ReferrerInfo(null, 0, 0, source);
        this.matchesGoogle = matchesGoogle;
        this.machesHuawei = matchesHuawei;
    }

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        context = TestUtils.createMockedContext();
        when(context.getPackageName()).thenReturn(packageName);
        referrerValidityChecker = new ReferrerValidityChecker(context, packageManager, selfReporter);
    }

    @Test
    public void googleInstaller() {
        when(packageManager.getInstallerPackageName(context, packageName)).thenReturn(GOOGLE_INSTALLER);
        assertThat(referrerValidityChecker.doesInstallerMatchReferrer(referrerInfo)).isEqualTo(matchesGoogle);
    }

    @Test
    public void huaweiInstaller() {
        when(packageManager.getInstallerPackageName(context, packageName)).thenReturn(HUAWEI_INSTALLER);
        assertThat(referrerValidityChecker.doesInstallerMatchReferrer(referrerInfo)).isEqualTo(machesHuawei);
    }

    @Test
    public void strangeInstaller() {
        when(packageManager.getInstallerPackageName(context, packageName)).thenReturn("bad installer");
        assertThat(referrerValidityChecker.doesInstallerMatchReferrer(referrerInfo)).isFalse();
    }

    @Test
    public void nullInstaller() {
        when(packageManager.getInstallerPackageName(context, packageName)).thenReturn(null);
        assertThat(referrerValidityChecker.doesInstallerMatchReferrer(referrerInfo)).isFalse();
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/referrer/service/ReferrerValidityCheckerDoesInstallerMatchReferrerTest.kt
package io.appmetrica.analytics.impl.referrer.service

import io.appmetrica.analytics.coreutils.internal.services.SafePackageManager
import io.appmetrica.analytics.impl.IReporterExtended
import io.appmetrica.analytics.impl.referrer.common.ReferrerInfo
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.ContextRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

@RunWith(Parameterized::class)
internal class ReferrerValidityCheckerDoesInstallerMatchReferrerTest(
    source: ReferrerInfo.Source,
    private val matchesGoogle: Boolean,
    private val matchesHuawei: Boolean
) : CommonTest() {

    private val packageManager: SafePackageManager = mock()
    private val selfReporter: IReporterExtended = mock()
    private val referrerInfo = ReferrerInfo("", 0, 0, source)

    @get:Rule
    val contextRule = ContextRule()
    val context by contextRule

    private val referrerValidityChecker by setUp { ReferrerValidityChecker(context, packageManager, selfReporter) }

    @Test
    fun googleInstaller() {
        whenever(packageManager.getInstallerPackageName(context, context.packageName))
            .thenReturn(GOOGLE_INSTALLER)
        assertThat(referrerValidityChecker.doesInstallerMatchReferrer(referrerInfo)).isEqualTo(matchesGoogle)
    }

    @Test
    fun huaweiInstaller() {
        whenever(packageManager.getInstallerPackageName(context, context.packageName))
            .thenReturn(HUAWEI_INSTALLER)
        assertThat(referrerValidityChecker.doesInstallerMatchReferrer(referrerInfo)).isEqualTo(matchesHuawei)
    }

    @Test
    fun strangeInstaller() {
        whenever(packageManager.getInstallerPackageName(context, context.packageName))
            .thenReturn("bad installer")
        assertThat(referrerValidityChecker.doesInstallerMatchReferrer(referrerInfo)).isFalse()
    }

    @Test
    fun nullInstaller() {
        whenever(packageManager.getInstallerPackageName(context, context.packageName)).thenReturn(null)
        assertThat(referrerValidityChecker.doesInstallerMatchReferrer(referrerInfo)).isFalse()
    }

    companion object {

        @JvmStatic
        @Parameterized.Parameters(name = "{0}")
        fun data(): Collection<Array<Any?>> = listOf(
            arrayOf(ReferrerInfo.Source.UNKNOWN, false, false),
            arrayOf(ReferrerInfo.Source.HMS, false, true),
            arrayOf(ReferrerInfo.Source.GP, true, false)
        )

        private const val GOOGLE_INSTALLER = "com.android.vending"
        private const val HUAWEI_INSTALLER = "com.huawei.appmarket"
    }
}
